Add ConstBuf

AntonReinhard opened the pull request:
I've started implementing a constant buffer as discussed in #1729 
This constant buffer type can be constructed by converting implicitly from a normal/non-const buffer type (to allow passing to functions for example), or explicitly through a `makeConstBuf` trait. The constant version then only allows retrieving constant pointers, which prevents mutation of the underlying data of the buffer. The buffer also shares ownership, so one could create a buffer, fill it with data, then convert the buffer to a constant buffer when no more mutation is necessary.

For now I have a version drafted for the CPU buffer only and would like some feedback.

AntonReinhard committed:
Add ConstCpuBuf and preliminary tests

AntonReinhard commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,119 @@
+    STATIC_REQUIRE(std::convertible_to<TBuf, TBuf>);
+    STATIC_REQUIRE(std::convertible_to<TBuf, TCBuf>);
+
+    // check that Constant buffer cannot give a non const native ptr
+    // STATIC_REQUIRE(buftest::onlyConstNativePtr<TCBuf>);
+    // getPtrNative(c_buf)* = 0.f;  // <- this does not compile, as desired
+    // however, the static require above doesn't work and seems to get float* instead of float const*
```
I don't understand what is going wrong here, if anyone has an idea I'd appreciate some help. The concept `onlyConstNativePtr` concept is defined above lines 68-71.

mehmetyusufoglu replied:
Is it possible to use more direct type checking like the ones below . There are decltype and auto variables it is not easy to follow. 
```

    static_assert(
        std::is_same_v<decltype(alpaka::getPtrNative(buftest::allocConstBuf<Dim, Elem, Idx>(dev, extent))), Elem const*>,
        "getPtrNative for buftest::allocConstBuf<Dim, Elem, Idx>(dev, extent) must return Elem const*");

    static_assert(
        std::is_same_v<decltype(alpaka::getPtrNative(alpaka::allocBuf<Elem, Idx>(dev, extent))), Elem const*>,
        "getPtrNative for alpaka::allocBuf must return Elem const*");

    static_assert(
        !std::is_same_v<decltype(alpaka::getPtrNative(buftest::allocConstBuf<Dim, Elem, Idx>(dev, extent))), Elem *>,
        "getPtrNative for buftest::allocConstBuf<Dim, Elem, Idx>(dev, extent) must NOT return Elem *");
```

SimeonEhrig replied:
You need to remove the `const` modifier of `ConstBufCpu` type:
```c++
    template<typename TBuf>
    concept onlyConstNativePtr = requires(TBuf t) {
        {
            alpaka::getPtrNative(t)
        } -> std::same_as<alpaka::Elem<std::remove_const_t<TBuf>> const*>;
    };
```

`alpaka::Elem` removes only [volatile](https://github.com/alpaka-group/alpaka/blob/1793a0760e57bd3cca2b71170507ae1c26b5c0dd/include/alpaka/elem/Traits.hpp#L21) and you didn't specialized the type trait `ElemType` for the const version. But I have no idea, if this is design or a bug.

fwyzard commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,119 @@
+
+#if defined(ALPAKA_ACC_GPU_CUDA_ENABLED)
+    template<typename TDim, typename TElem, typename TIdx, typename TExtent>
+    auto allocBuf(alpaka::DevCudaRt dev, TExtent extent) -> alpaka::ConstBufCudaRt<TElem, TDim, TIdx>
```
Shouldn't this be `allocConstBuf` ?
Same for the functions below.

AntonReinhard replied:
Yes, but I haven't implemented the Buffer for the other Accelerators yet anyway. I wanted to make a draft with the CPU Buffer first before I go and do it for all the other devices too.

psychocoderHPC commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,119 @@
+    template<typename TDim, typename TElem, typename TIdx, typename TExtent>
+    auto allocBuf(alpaka::DevGpuSyclIntel dev, TExtent extent) -> alpaka::ConstBufGpuSyclIntel<TElem, TDim, TIdx>
+    {
+        return alpaka::makeConstBuf(alpaka::allocBuf<TElem, TIdx>(dev, extent));
```
I have currently problems following the PR and I am not sure anymore what this PR is trying to solve, maybe we should discuss this in our developer meeting.
I thought the `ConstBuff` should convert an exiting buffer into a nonmutable buffer, which makes IMO sense.
In this line we create a constant buffer that is not mutable, I am asking myself how to fill  nonmutable buffer with data :thinking:

AntonReinhard replied:
As I also explained in the meeting, this `allocBuf` function is only used for testing where I want to check that it is impossible to mute the ConstBuf's buffer. Therefore I don't care what data is inside.
The `makeConstBuf` function is simply a convenient explicit cast, creating a constant buffer from the mutable buffer. It just calls the copy constructor, but calling the copy constructor yourself would be tedious because you'd need to also provide all the correct template arguments for the resulting `ConstBuf`.

SimeonEhrig commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,119 @@
+
+    // check that Constant buffer cannot give a non const native ptr
+    // STATIC_REQUIRE(buftest::onlyConstNativePtr<TCBuf>);
+    // getPtrNative(c_buf)* = 0.f;  // <- this does not compile, as desired
```
```suggestion
    // *getPtrNative(c_buf) = 0.f;  // <- this does not compile, as desired
```

Typo and syntax error.

SimeonEhrig commented on include/alpaka/mem/buf/cpu/ConstBufCpu.hpp:
```diff
@@ -0,0 +1,129 @@
+{
+    //! The constant CPU memory buffer.
+    template<typename TElem, typename TDim, typename TIdx>
+    class ConstBufCpu : public internal::ViewAccessOps<ConstBufCpu<TElem, TDim, TIdx>>
```
Like @fwyzard suggest, maybe it is a better idea to inherit from `BufCpu` and only specialize the [GetPtrNative](https://github.com/AntonReinhard/alpaka/blob/59d22208e7caa1a5c779e66556e9e99f8463e798/include/alpaka/mem/buf/BufCpu.hpp#L153) trait to save code.

AntonReinhard committed:
WIP

AntonReinhard commented:
I'm redrafting this to try to prevent too much code duplication.

AntonReinhard committed:
Add GenericMutableBuf and make CpuBuf const by default

AntonReinhard committed:
Use flatter dispatch implementation from Buf to ConstBuf

AntonReinhard committed:
Format and end of day

AntonReinhard commented:
In order to make my solution truly work generically, I'd have to change the implementation of the `BufGenericSycl` and `BufUniformCudaHipRt` and make them work like the `BufCpu`, i.e., hide it away in an extra Impl class. The pointers (currently `m_spBufCpuImpl` for the CPU impl, `m_spMem` for Sycl and CUDA/hip) are public fields in the classes, so if I change these, they will potentially break compatability. I'm not sure why these are public in the first place though.

Is that a problem?

AntonReinhard committed:
Make impl pointer in BufCpu private and rename generic implementation

AntonReinhard committed:
Making sycl constbuf WIP

AntonReinhard committed:
Restructure WIP

AntonReinhard committed:
Fix public -> private in CpuBuf

AntonReinhard committed:
WIP Sycl Buffers

AntonReinhard committed:
Fix sycl constbuf implementation

AntonReinhard committed:
Apply constbuf changes to UniformCudaHipRt buffer

AntonReinhard committed:
Fix headercheck test

AntonReinhard committed:
Apply format patch

AntonReinhard committed:
Fix checkHeaders for uniform cuda/hip buffer

SimeonEhrig commented:
@AntonReinhard Where is the `makeConstBuf` function? In the unit test, I found preprocessor cascade, which is not suitable for the user API. If I should review an intermediate state, the `makeConstBuf` function is still useful as entry point for the review.

AntonReinhard commented:
> @AntonReinhard Where is the `makeConstBuf` function? In the unit test, I found preprocessor cascade, which is not suitable for the user API. If I should review an intermediate state, the `makeConstBuf` function is still useful as entry point for the review.

I don't currently have a `makeConstBuf` anymore. I could add it again if it is deemed more user friendly. Currently, you simply cast an existing buffer into the respective `ConstBuf` version. The implementations are the constructors in the `ConstBuf<CPU/SYCL/CUDA/HIP>` files.

SimeonEhrig commented:
> > @AntonReinhard Where is the `makeConstBuf` function? In the unit test, I found preprocessor cascade, which is not suitable for the user API. If I should review an intermediate state, the `makeConstBuf` function is still useful as entry point for the review.
> 
> I don't currently have a `makeConstBuf` anymore. I could add it again if it is deemed more user friendly. Currently, you simply cast an existing buffer into the respective `ConstBuf` version. The implementations are the constructors in the `ConstBuf<CPU/SYCL/CUDA/HIP>` files.

Well, it is a little bit hard to review a functionally, if there is no code to demonstrate how to use it :grimacing: But I can play around with casting the buffer.
Actual I'm a little bit surprised, that you don't wrote an test and an example which uses the user API. If I start implement a function, I write the user API in a early stage to check, if the API is doable and my approach is working like thought. But if you get it manage it without it, it is fine. I think it is only uncommon.

AntonReinhard commented:
> > > @AntonReinhard Where is the `makeConstBuf` function? In the unit test, I found preprocessor cascade, which is not suitable for the user API. If I should review an intermediate state, the `makeConstBuf` function is still useful as entry point for the review.
> > 
> > 
> > I don't currently have a `makeConstBuf` anymore. I could add it again if it is deemed more user friendly. Currently, you simply cast an existing buffer into the respective `ConstBuf` version. The implementations are the constructors in the `ConstBuf<CPU/SYCL/CUDA/HIP>` files.
> 
> Well, it is a little bit hard to review a functionally, if there is no code to demonstrate how to use it 😬 But I can play around with casting the buffer. Actual I'm a little bit surprised, that you don't wrote an test and an example which uses the user API. If I start implement a function, I write the user API in a early stage to check, if the API is doable and my approach is working like thought. But if you get it manage it without it, it is fine. I think it is only uncommon.

The test in `ConstBufTest.cpp` does test the user API. It allocates a "normal" mutable buffer using the `alpaka::allocBuf` trait, then converts it immediately to the constant version of it. 

I should add more tests for this functionality, maybe I can find another test that would benefit from being switched to a `ConstBuf` usage somewhere, to have a more real use case.

AntonReinhard committed:
Add makeConstBuf trait and use constbufs in matmul test

AntonReinhard commented:
I added the `makeConstBuf` function and trait back. It now simply takes a buffer (mutable or constant) and returns a constant one with the same memory set (and no copy performed), but only read access allowed.
I used this in the matmul test, where the A and B matrices don't need write access within the kernel. They are now created, then filled with their contents, and then made into constant versions, which are then used for the rest of the test.

AntonReinhard commented:
I also split the buffer definitions and implementations in multiple files and tried to make it as consistent as possible across the different implementations. For every implementation (cpu, sycl, uniformcudahip) there are now (here given as example for the CPU ones):
- CpuBufImpl.hpp: The definition of the class storing the dimension, extent, etc. of the actual buffer.
- MutBufCpu.hpp: The definition of the class for a mutable buffer.
- ConstBufCpu.hpp: The definition of the class for a constant buffer.
- BufCpu.hpp: The main file importing the `MutBuf` and `ConstBuf` versions, their traits, `Copy.hpp` and `Set.hpp`, and making an alias from the `MutBufCpu` template to the `BufCpu` name in the same style as before this PR.
- traits/ConstBufCpuTraits.hpp: Implementation of the traits concerning the constant buffer.
- traits/MutBufCpuTraits.hpp: Implementation of the traits concerning the mutable buffer.
- specialization/*.hpp: Specializations of the constant and mutable buffer names for specific devices, in the case of sycl and uniformcudahip.

SimeonEhrig commented on include/alpaka/mem/buf/Traits.hpp:
```diff
@@ -189,4 +194,16 @@ namespace alpaka
+    {
+        return trait::MakeConstBuf<TBuf>::makeConstBuf(buf);
+    }
+
```
Is it also possible to implement an `alpaka::ConstBuf` like `alpaka::Buf` it is implemented in this file?

https://github.com/alpaka-group/alpaka/blob/d3286539ba4a3037267b1cfc86e6cb300b4ff4af/include/alpaka/mem/buf/Traits.hpp#L56-L57

The definition is used in the examples, like here
https://github.com/alpaka-group/alpaka/blob/e906b1c0c449ebab02b45db6475246c972251aad/example/monteCarloIntegration/src/monteCarloIntegration.cpp#L97-L98

AntonReinhard replied:
That should be easy to add :+1:

AntonReinhard replied:
Done in ba70daa61920643357f1fe0a930eb2e3720067eb

SimeonEhrig commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,111 @@
+        return alpaka::makeConstBuf(alpaka::allocBuf<TElem, TIdx>(dev, extent));
+    }
+
+#if defined(ALPAKA_ACC_GPU_CUDA_ENABLED)
```
I don't think, we need the `#if`-else cascade anymore. If a backend does not implement the required traits, we get a compiler error.
You should test against the generic interface. The test automatically iterate over all enable backends. The rest is done by the CI.

AntonReinhard replied:
Done in 501d2b1a765f26e65c785584fe25cdb321152fcb

SimeonEhrig commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,111 @@
+/* Copyright 2025 Anton Reinhard
```
Idea for an test: Can you please write a test, which creates an `alpaka::Buf`, creates `alpaka::ConstBuf` from it and the `alpaka::Buf` should run of scope. So the `alpaka::ConstBuf` should still point to valid memory and `alpaka::Buf` should not delete it.

AntonReinhard replied:
That's a good idea :+1:

AntonReinhard replied:
Actually, the ConstBufTest already does this.
```cpp
template<typename TDev, typename TDim, typename TElem, typename TIdx, typename TExtent>
auto allocConstBuf(TDev dev, TExtent extent) -> alpaka::ConstBuf<TDev, TElem, TDim, TIdx>
{
    return alpaka::makeConstBuf(alpaka::allocBuf<TElem, TIdx>(dev, extent));
}
```
This function creates a normal buffer, then creates a new constant buffer from it and returns only that. The temporary originally allocated buffer goes out of scope there.

fwyzard commented on include/alpaka/mem/buf/cpu/CpuBuf.hpp:
The alias is called `BufCpu`, so the file should be called `BufCpu.hpp`, not `CpuBuf.hpp`.

AntonReinhard replied:
There were some inconsistencies in the naming of classes and files beforehand, I mostly tried to keep the inconsistencies "consistently". I'll try to rename things to make them make more sense.

AntonReinhard replied:
Done in ef24409ef8de95123204c9be82b4d48a568eebea

SimeonEhrig commented:
We need a VC. I don't understand your implementation. The type of `alpaka::ConstBufCpu` is understandable:
`alpaka::ConstBufCpu<unsigned int, std::integral_constant<unsigned long, 1ul>, unsigned long>`

And you explained me, that `alpaka::MutBufCpu` "inherits" from `alpaka::ConstBufCpu` but I don't the mechanism. This is the type of `alpaka::MutBufCpu`:
`alpaka::MutBufCpu<alpaka::ConstBufCpu, alpaka::detail::BufCpuImpl<unsigned int, std::integral_constant<unsigned long, 1ul>, unsigned long>, alpaka::DevCpu, unsigned int, std::integral_constant<unsigned long, 1ul>, unsigned long>`

fwyzard commented on include/alpaka/mem/buf/cpu/CpuBuf.hpp:
```diff
@@ -0,0 +1,17 @@
+namespace alpaka
+{
+    template<typename TElem, typename TDim, typename TIdx>
+    using BufCpu = MutBufCpu<ConstBufCpu, alpaka::detail::BufCpuImpl<TElem, TDim, TIdx>, DevCpu, TElem, TDim, TIdx>;
```
In fact, what is the point of having a `MutBufCpu` class and a `BufCpu` alias ?
Why not simply have a `BufCpu` class ?

AntonReinhard replied:
The point was to hide the additional template arguments I introduced from users of the class. But I can try to get rid of them by deriving them inside the class directly, I think it should be possible.

fwyzard replied:
Mhm, I see.

SimeonEhrig replied:
We should discuss this via VC or directly. I still think, we can make it easier. For me, I don't understand why we need the template parameter `ConstBufCpu` **and** `alpaka::detail::BufCpuImpl` **and** `DevCpu`.

I think, if I understand the code, we can find a more understandable solution, maybe by inheritance, maybe with traits and concepts or something else.

AntonReinhard replied:
Yes I already noticed they are not needed anymore at all, it was leftover from my previous more generic attempt at an implementation. Please wait a bit until I can revise what I have, then we can talk about that.

AntonReinhard replied:
Please take a look at ef24409ef8de95123204c9be82b4d48a568eebea

fwyzard commented on include/alpaka/mem/buf/cpu/CpuBufImpl.hpp:
`BufCpuImpl` name vs `CpuBufImpl.hpp`

AntonReinhard replied:
Done in ef24409ef8de95123204c9be82b4d48a568eebea

fwyzard reviewed (changes requested):
Please do not introduce new names like `MutBufUniformCudaHipRt` and `MutBufCpu`, and then aliases for the old names.

AntonReinhard committed:
Remove unneeded template arguments and explicit MutBuf versions

AntonReinhard committed:
Formatting

AntonReinhard committed:
Add ConstBuf trait and use it in ConstBuf test

AntonReinhard committed:
Fix ConstBufTest

AntonReinhard commented:
> We need a VC. I don't understand your implementation. The type of `alpaka::ConstBufCpu` is understandable: `alpaka::ConstBufCpu<unsigned int, std::integral_constant<unsigned long, 1ul>, unsigned long>`

In ef24409ef8de95123204c9be82b4d48a568eebea, I removed the extra template arguments from the mutable buffer types, so they should look roughly the same as before the PR. They were leftovers from my previous attempt at an implementation that I reworked.

> And you explained me, that `alpaka::MutBufCpu` "inherits" from `alpaka::ConstBufCpu` but I don't the mechanism. This is the type of `alpaka::MutBufCpu`: `alpaka::MutBufCpu<alpaka::ConstBufCpu, alpaka::detail::BufCpuImpl<unsigned int, std::integral_constant<unsigned long, 1ul>, unsigned long>, alpaka::DevCpu, unsigned int, std::integral_constant<unsigned long, 1ul>, unsigned long>`

The mutable buffer types don't actually inherit from the const buffer types anymore. They only use the constant buffers' implementations of all the traits by casting themselves to a const buffer and then calling the trait on that.
For example:
```cpp
    //! The BufCpu width get trait specialization.
    template<typename TElem, typename TDim, typename TIdx>
    struct GetExtents<BufCpu<TElem, TDim, TIdx>>
    {
        ALPAKA_FN_HOST auto operator()(BufCpu<TElem, TDim, TIdx> const& buf)
        {
            return GetExtents<ConstBufCpu<TElem, TDim, TIdx>>{}(ConstBufCpu<TElem, TDim, TIdx>{buf});
        } //                  ^ Calling the const buf trait     ^ Casting to a const buf
    };
```

AntonReinhard commented:
Now that I think about it, the casts might not be able to be compiled away by the compiler because they copy a `shared_ptr` internally. Maybe they should not cast and instead just copy the implementation as code...

SimeonEhrig commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,74 @@
+} // namespace buftest
+
+template<typename TAcc>
+static auto testConstBuffer(alpaka::Vec<alpaka::Dim<TAcc>, alpaka::Idx<TAcc>> const& extent) -> void
```
Can you please add also an test, which checks the return type of the access operator of a const and non const buffer. Therefor you need to allocate an 1D buffer.

```c++
static_assert(std::is_same_v<decltype(buf[0]), int&>);
```

AntonReinhard replied:
I added it in dde2ed4f0e99a39d865aeca36d627b1f9b842052

SimeonEhrig commented:
The current implementation looks much better. Please add the test case, which I suggested. Else @fwyzard will do the final review.

AntonReinhard committed:
Move buffer allocation traits

AntonReinhard committed:
Duplicate small buffer trait implementations to save shared_ptr copies

AntonReinhard committed:
Rename BufImpl members consistently

AntonReinhard committed:
Fix renaming

AntonReinhard committed:
Add const buf lifetime test

AntonReinhard committed:
Fix ConstBufTest

AntonReinhard committed:
Add explicit [] type check for constbuf

fwyzard commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,136 @@
+    std::vector<Elem> dataHost(static_cast<std::size_t>(getExtentProduct(extent)), static_cast<Elem>(1));
+    auto bufHost = alpaka::createView(devHost, dataHost.data(), extent);
+
+    auto copiedBack = alpaka::allocBuf<Elem, Idx>(devHost, extent);
```
why `copiedBack` ?
if I understand correctly the test, the data is always on CPU:
  - in `dataHost`
  - in `bufHost`
  - in `copiedBack`

AntonReinhard replied:
In the next line, I `memcpy` the data from the `c_buf`, which lives on the accelerator, back to the CPU to be able to compare it with the expected results.

fwyzard commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,136 @@
+}
+
+template<typename TAcc>
+static auto testConstBufLifetime(alpaka::Vec<alpaka::Dim<TAcc>, alpaka::Idx<TAcc>> const& extent) -> void
```
I don't think this test actually does any tests of the buffers lifetime.
Maybe the implementation should be different ?
Or maybe just the name ?

AntonReinhard replied:
In `initConstBuf`, a normal buffer is allocated, initialized, then converted to a constant buffer. The originally allocated buffer then runs out of scope. So the intention is that this tests that the returned constant buffer still owns the memory and it didn't get deleted when the original non-constant buffer runs out of scope.

fwyzard replied:
Ah, OK, it's kind of hidden form the test itself, but I see what you mean.

fwyzard commented on test/unit/mem/buf/src/ConstBufTest.cpp:
```diff
@@ -0,0 +1,136 @@
+        alpaka::wait(queueAcc);
+
+        // make the buffer constant and return only that
+        auto const constBuf = alpaka::makeConstBuf(buf);
```
Would `alpaka::makeConstBuf(std::move(buf))` work ?
Or does it break the type deduction in the trait ?

AntonReinhard replied:
It seems to work :+1:

fwyzard commented on include/alpaka/mem/buf/cpu/ConstBufCpu.hpp:
```diff
@@ -0,0 +1,58 @@
+        {
+        }
+
+        ALPAKA_FN_HOST ConstBufCpu(BufCpu<TElem, TDim, TIdx> const& buf) : m_spBufImpl{buf.m_spBufImpl}
+        {
+        }
```
Can you add an overload
```c++
        ALPAKA_FN_HOST ConstBufCpu(BufCpu<TElem, TDim, TIdx> && buf) : m_spBufImpl{std::move(buf.m_spBufImpl)} {}
```
?

fwyzard replied:
(also for the other buffer types)

AntonReinhard replied:
Added in fbb8325e3379c28a5da44a557eccaea75d8dc32d

fwyzard commented on include/alpaka/mem/buf/uniformCudaHip/traits/ConstBufUniformCudaHipRtTraits.hpp:
```diff
@@ -0,0 +1,205 @@
+
+            return BufCpu<TElem, TDim, TIdx>(host, memPtr, std::move(deleter), extent);
+        }
+    };
```
Move the allocation to the non-const buffer traits ?

AntonReinhard replied:
Forgot this one... Done in fbb8325e3379c28a5da44a557eccaea75d8dc32d

fwyzard commented on include/alpaka/mem/buf/sycl/traits/ConstBufGenericSyclTraits.hpp:
```diff
@@ -0,0 +1,143 @@
+/* Copyright 2024 Jan Stephan, Luca Ferragina, Aurora Perego, Andrea Bocci
```
Add your name ?

AntonReinhard replied:
Done in fbb8325e3379c28a5da44a557eccaea75d8dc32d

fwyzard commented on include/alpaka/mem/buf/uniformCudaHip/BufUniformCudaHipRt.hpp:
```diff
@@ -0,0 +1,54 @@
+
+} // namespace alpaka
+
+#    include "alpaka/mem/buf/uniformCudaHip/Copy.hpp"
+#    include "alpaka/mem/buf/uniformCudaHip/Set.hpp"
+#    include "alpaka/mem/buf/uniformCudaHip/traits/BufUniformCudaHipRtTraits.hpp"
+#    include "alpaka/mem/buf/uniformCudaHip/traits/ConstBufUniformCudaHipRtTraits.hpp"
```
I find it confusing that we have these includes in the mutable buffer's header, and not in the const buffer's header.
(I understand that the latter includes the former, but it seems more of an accident than a strict requirement).

Do they need to be include after the definition of `BufUniformCudaHipRt` (but not after the definition of `ConstBufUniformCudaHipRt`) ?

~~Would it work to move them at the end of `BufUniformCudaHipRtImpl.hpp` ?~~

No, that doesn't seem the right place either...

fwyzard replied:
OK, if there is no better place, I guess they can stay here.

AntonReinhard committed:
Review fixes

fwyzard commented on include/alpaka/mem/buf/uniformCudaHip/traits/ConstBufUniformCudaHipRtTraits.hpp:
```diff
@@ -0,0 +1,180 @@
+    template<typename TApi>
+    struct HasMappedBufSupport<PlatformUniformCudaHipRt<TApi>> : public std::true_type
+    {
+    };
```
Sorry, I missed them earlier: these should also go to the mutable buffer, since they describe what allocation methods are supported.

AntonReinhard replied:
Moved in 78cf911ebb9e15e8c45de565f46abb7fff0cf3af

fwyzard commented:
OK, I have two last general comments...

Comment #1: can you make sure that one can call `alpaka::makeConstBuf(std::move(buf))` and have the move operation propagate to all implementations, so that there is no increment/decrement of the shared pointer counter ?

Comment #2: I'm still convinced that this could be implemented making `Buf` inherit from `ConstBuf`, which could avoid the cast and the temporary when a `Buf` is passed to a function that takes a `ConstBuf const&` - but we can go ahead and merge this PR, and explore this in a follow up.

AntonReinhard committed:
Review comments

AntonReinhard committed:
Make makeConstBuf able to std::move properly

SimeonEhrig commented:
There is a concept to check, if a object is moveable. https://en.cppreference.com/w/cpp/concepts/movable
```c++
// e.g.
STATIC_REQUIRED(std::moveable<ConstBuf<...>>);
```

AntonReinhard committed:
Add movability test

AntonReinhard commented:
> There is a concept to check, if a object is moveable. https://en.cppreference.com/w/cpp/concepts/movable
> 
> ```c++
> // e.g.
> STATIC_REQUIRED(std::moveable<ConstBuf<...>>);
> ```

Thanks for that hint, I added a test for this as well.

Otherwise I also agree with @fwyzard 's comment 2, I would prefer to merge this now and I can look at the inheritance possibility again in a separate PR.
